R/mock-object.R

```diff
@@ -0,0 +1,174 @@
+
+#' Create a mock object.
+#'
+#' Mock object's primary use is to record calls that are made on the
+#' mocked function.
+#'
+#' Optionally values/expressions can be passed via \code{...} for the
+#' mock object to return them upon subsequent calls. Expressions are
+#' evaluated in environment \code{envir} before being returned. If no
+#' value is passed in \code{...} then \code{NULL} is returned.
+#'
+#' Passing an expression or a function call via \code{...} is also a
+#' way to implement side effects: keep track of the state of code
+#' under testing, throw an exception when a condition is met, etc.
+#'
+#' @param ... Values returned upon subsequent calls.
+#' @param cycle Whether to cycle over the return values. If \code{FALSE},
+#'        will fail if called too many times.
+#' @param envir Where to evaluate the expressions being returned.
+#' @return A mock object.
+#'
+#' @export
+#' @examples
+#' m <- mock(1)
+#' with_mock(summary = m, {
+#'   expect_equal(summary(iris), 1)
+#'   expect_length(m, 1)
+#'   expect_call(m, 1, summary(iris))
+#'   expect_args(m, 1, iris)
+#' })
+#'
+mock <- function (..., cycle = FALSE, envir = parent.frame()) {
+  stopifnot(is.environment(envir))
+
+  return_values     <- eval(substitute(alist(...)))
```

Comment:
Oh because you're capturing them unevaluated, assuming that they're coming from `envir`. That's not a great assumption because each element of `...` could actually come from a different environment. I think this will have to wait until the next release where testthat will depend on lazyeval, and you can do this correctly.